File: webscrambles/src/net/gnehzr/tnoodle/server/webscrambles/OrderedScrambles.java
package net.gnehzr.tnoodle.server.webscrambles;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.Days;

import net.lingala.zip4j.exception.ZipException;
import net.lingala.zip4j.io.ZipOutputStream;
import net.lingala.zip4j.model.ZipParameters;

import com.google.gson.JsonElement;

import com.itextpdf.text.DocumentException;

import static net.gnehzr.tnoodle.server.webscrambles.Utils.toFileSafeString;
import static net.gnehzr.tnoodle.utils.GwtSafeUtils.azzert;

public class OrderedScrambles {
    
    // TODO see https://github.com/thewca/tnoodle/issues/400

    public static void generateOrderedScrambles(String globalTitle, Date generationDate, ZipOutputStream zipOut, ZipParameters parameters, WCIFHelper wcifHelper) throws DocumentException, IOException, ZipException {
        
        if (wcifHelper == null || wcifHelper.getSchedule() == null) {
            return;
        }
        
        azzert(wcifHelper.getAllScrambleRequests() != null, "There should be scramble requests.");
        
        // allActivities hold activities without repetition.
        // A ScrambleRequest is added to AllScramblesOrdered if its activity is not on allActivities.
        // This is needed because the same activity may happen on different rooms.
        ArrayList<ScrambleRequest> allScramblesOrdered = new ArrayList<ScrambleRequest>();
        ArrayList<String> allActivities = new ArrayList<String>();

        boolean hasMultipleDays = wcifHelper.hasMultipleDays();
        boolean hasMultipleVenues = wcifHelper.hasMultipleVenues();
        
        // We consider the competition start date as the earlier activity from the schedule.
        // This prevents miscalculation of dates for multiple timezones.
        String competitionStartString = wcifHelper.getEarlierActivityString();
        
        for (JsonElement venue : wcifHelper.getVenues()) {
            String venueName = wcifHelper.getSafeVenueName(venue);
                        
            DateTimeZone timezone = wcifHelper.getTimeZone(venue);
            DateTime competitionStartDate = new DateTime(competitionStartString, timezone);
            
            boolean hasMultipleRooms = wcifHelper.hasMultipleRooms(venue);

            for (JsonElement room : wcifHelper.getRooms(venue)) {

                ArrayList<Integer> dayList = new ArrayList<Integer>();
                ArrayList<ArrayList<ScrambleRequest>> scrambleRequestListByDay = new ArrayList<ArrayList<ScrambleRequest>>();
                
                String roomName = wcifHelper.getSafeRoomName(room);
                                
                for (JsonElement activity : wcifHelper.getActivities(room)) {
                    String activityCode = wcifHelper.getActivityCode(activity);
                    
                    int indexAllActivities = allActivities.indexOf(activityCode);
                    if (indexAllActivities < 0) {
                        allActivities.add(activityCode);
                    }

                    for (ScrambleRequest scrambleRequest : wcifHelper.getScrambleRequests(activityCode)) {

                        DateTime activityStartTime = wcifHelper.getActivityStartTime(activity, timezone);

                        int activityDay = Days.daysBetween(competitionStartDate.withTimeAtStartOfDay(), activityStartTime.withTimeAtStartOfDay()).getDays()+1;
                        
                        if (!dayList.contains(activityDay)) {
                            dayList.add(activityDay);
                            scrambleRequestListByDay.add(new ArrayList<ScrambleRequest>());
                        }

                        int index = dayList.indexOf(activityDay);
                        scrambleRequest.roundStartTime = activityStartTime;
                        scrambleRequestListByDay.get(index).add(scrambleRequest);
                        
                        // A case to be considered here is the scenario where the competition
                        // for whatever reason, has the same activity happening on different start time
                        // If this is true, we currently have no control on which one is going to be considered.
                        // This is not serious because it makes no sense to repeat activity on the schedule.
                        if (indexAllActivities < 0) {
                            allScramblesOrdered.add(scrambleRequest);
                        }
                    }
                }

                // hasMultipleDays gets a variable assigned on the competition creation using the website's form.
                // Online schedule fit to it and the user should not be able to put events outside it, but we double check here.
                // The next assignment fix possible mistakes (eg. a competition is assigned with 1 day, but events are spread among 2 days).
                hasMultipleDays = hasMultipleDays || dayList.size() > 1;

                for (int index = 0; index<dayList.size(); index++) {
                    if (scrambleRequestListByDay.get(index).size()>0) {
                        
                        Collections.sort(scrambleRequestListByDay.get(index));
                        
                        String pdfFileName = "Printing/Ordered Scrambles/";

                        if (hasMultipleVenues) {
                            pdfFileName += venueName+"/";
                        }

                        if (hasMultipleDays) {
                            pdfFileName += "Day "+dayList.get(index)+"/";
                        }

                        pdfFileName += "Ordered Scrambles";
                        
                        // In addition to different folders, we stamp venue, day and room in the PDF's name
                        // to prevent different files with the same name.
                        if (hasMultipleVenues) {
                            pdfFileName += " - " + venueName;
                        }
                        
                        if (hasMultipleDays) {
                            pdfFileName += " - Day "+dayList.get(index);
                        }
                        
                        if (hasMultipleRooms) {
                            pdfFileName += " - "+roomName;
                        }
                        pdfFileName += ".pdf";
                        
                        parameters.setFileNameInZip(pdfFileName);
                        zipOut.putNextEntry(null, parameters);
                        ScrambleRequest[] scrambleRequests = scrambleRequestListByDay.get(index).toArray(new ScrambleRequest[scrambleRequestListByDay.get(index).size()]);
                        ByteArrayOutputStream baos = ScrambleRequest.requestsToPdf(globalTitle, generationDate, scrambleRequests, null);
                        zipOut.write(baos.toByteArray());
                        zipOut.closeEntry();
                    }
                }
            }
        }
        
        // Generate all scrambles ordered
        Collections.sort(allScramblesOrdered);
        String safeGlobalTitle = toFileSafeString(globalTitle);
        String pdfFileName = "Printing/Ordered Scrambles/Ordered "+safeGlobalTitle+" - All Scrambles.pdf";

        parameters.setFileNameInZip(pdfFileName);
        zipOut.putNextEntry(null, parameters);
        ScrambleRequest[] scrambleRequests = allScramblesOrdered.toArray(new ScrambleRequest[allScramblesOrdered.size()]);
        ByteArrayOutputStream baos = ScrambleRequest.requestsToPdf(globalTitle, generationDate, scrambleRequests, null);
        zipOut.write(baos.toByteArray());
        zipOut.closeEntry();
    }
}

File: webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/wcif/OrderedScrambles.kt
package org.worldcubeassociation.tnoodle.server.webscrambles.wcif

import org.worldcubeassociation.tnoodle.server.webscrambles.ScrambleRequest
import org.worldcubeassociation.tnoodle.server.webscrambles.wcif.WCIFParser.atLocalStartOfDay
import org.worldcubeassociation.tnoodle.server.webscrambles.zip.FolderBuilder
import java.time.LocalDate
import java.time.Period

object OrderedScrambles {
    fun generateOrderedScrambles(builder: FolderBuilder, globalTitle: String?, generationDate: LocalDate, versionTag: String, wcifConfig: WCIFRequestBinding) {
        val wcifSchedule = wcifConfig.wcif.schedule

        if (wcifSchedule.venues.isEmpty()) {
            return
        }

        builder.folder("Ordered Scrambles") {
            orderedScrambles(globalTitle, generationDate, versionTag, wcifConfig)
        }
    }

    fun FolderBuilder.orderedScrambles(globalTitle: String?, generationDate: LocalDate, versionTag: String, wcifConfig: WCIFRequestBinding) {
        val wcifSchedule = wcifConfig.wcif.schedule

        val activityDays = wcifSchedule.activitiesWithLocalStartTimes
            .map { it.value.dayOfYear }
            .distinct()

        // hasMultipleDays gets a variable assigned on the competition creation using the website's form.
        // Online schedule fit to it and the user should not be able to put events outside it, but we double check here.
        // The next assignment fix possible mistakes (eg. a competition is assigned with 1 day, but events are spread among 2 days).
        val hasMultipleDays = wcifSchedule.hasMultipleDays || activityDays.size > 1
        val hasMultipleVenues = wcifSchedule.hasMultipleVenues

        // We consider the competition start date as the earlier activity from the schedule.
        // This prevents miscalculation of dates for multiple timezones.
        val competitionStartActivity = wcifSchedule.earliestActivity

        for (venue in wcifSchedule.venues) {
            val venueName = venue.fileSafeName
            val hasMultipleRooms = venue.hasMultipleRooms

            val timezone = venue.dateTimeZone
            val competitionStartDate = competitionStartActivity.getLocalStartTime(timezone)

            for (room in venue.rooms) {
                val roomName = room.fileSafeName

                val activitiesPerDay = room.activities
                    .groupBy {
                        Period.between(
                            competitionStartDate.atLocalStartOfDay(),
                            it.getLocalStartTime(timezone).atLocalStartOfDay()
                        ).days
                    }

                for ((nthDay, activities) in activitiesPerDay) {
                    val scrambles = wcifConfig.activityScrambleRequests
                        .filterKeys { it in activities }

                    val activitiesHaveScrambles = scrambles.values.any { it.isNotEmpty() }

                    if (activitiesHaveScrambles) {
                        val filenameDay = nthDay + 1

                        val parts = listOfNotNull(
                            "$venueName/".takeIf { hasMultipleVenues },
                            "Day $filenameDay/".takeIf { hasMultipleDays },
                            "Ordered Scrambles",
                            " - $venueName".takeIf { hasMultipleVenues },
                            " - Day $filenameDay".takeIf { hasMultipleDays },
                            " - $roomName".takeIf { hasMultipleRooms },
                            ".pdf"
                        )

                        if (hasMultipleVenues || hasMultipleDays || hasMultipleRooms) {
                            // In addition to different folders, we stamp venue, day and room in the PDF's name
                            // to prevent different files with the same name.
                            val pdfFileName = parts.joinToString("")

                            val sortedScrambles = scrambles.entries
                                .sortedBy { it.key.getLocalStartTime(timezone) }
                                .flatMap { it.value }

                            val sheet = ScrambleRequest.requestsToCompletePdf(globalTitle, generationDate, versionTag, sortedScrambles)
                            file(pdfFileName, sheet.render())
                        }
                    }
                }
            }
        }

        // Generate all scrambles ordered
        val allScramblesOrdered = wcifSchedule.activitiesWithLocalStartTimes.entries
            .sortedBy { it.value }
            .flatMap { wcifConfig.activityScrambleRequests[it.key].orEmpty() }
            .distinct()

        val completeOrderedPdf = ScrambleRequest.requestsToCompletePdf(globalTitle, generationDate, versionTag, allScramblesOrdered)

        file("Ordered $globalTitle - All Scrambles.pdf", completeOrderedPdf.render())
    }
}
